src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1277,287 +804,149 @@ where
-            let ingest_frontier = frontiers[0].frontier();
-            for (cap, (pid, offset)) in source_upper_updates.iter() {
-                if ingest_frontier.less_equal(cap) {
-                    continue;
-                }
+                    // We only want to retain non-empty batches
+                    untimestamped_batches.retain(|batch| !batch.is_empty());
 
-                trace!(
-                    "reclock({id}) {worker_id}/{worker_count}: \
-                    applying source_upper update ({:?}, {:?}) \
-                    global_source_upper {:?}",
-                    pid,
-                    offset,
-                    global_source_upper
-                );
+                    // And downgrade our capability to the reclocked source upper
 
-                global_source_upper.maybe_insert(pid.clone(), *offset);
-            }
-            source_upper_updates.retain(|(cap, _update)| ingest_frontier.less_equal(cap));
-
-            // It can happen that our view of the global source_upper is not yet
-            // up to date with what the ReclockOperator thinks. We will
-            // evantually learn about an up-to-date frontier in a future
-            // invocation.
-            //
-            // Note that this even if the `global_source_upper` has advanced past
-            // the remap input, this holds back capability downgrades until
-            // they match.
-            //
-            // We "meet" the _source upper_ and the _lower_ of batches that we
-            // have stashed. We cannot downgrade our capability beyond the
-            // frontier that corresponds to that lower, otherwise we would not
-            // be able to emit batches once we succesfully reclock them.
-            //
-            // Note that `bounded` is not commutative, but we want to ensure that
-            // if the `source_upper` has some offset for a partition not yet in
-            // the lower, we consider that partition when reclocking that frontier,
-            // as that partition's timestamp may be behind the timestamps
-            // of the partitions in the current batch lower.
-            let global_batch_lower = batch_capabilities.frontier();
-            let bounded_source_upper = global_source_upper.bounded(&global_batch_lower);
-
-            if let Ok(new_ts_upper) = timestamper.reclock_frontier(
-                FromTime::from_compat_frontier(bounded_source_upper.clone()).borrow(),
-            ) {
-                tracing::trace!(
-                    "reclock({id}) {worker_id}/{worker_count}: \
-                        global_batch_lower: {:?}, global_source_upper: \
-                        {:?}, bounded_source_upper: {:?}",
-                    global_batch_lower,
-                    global_source_upper,
-                    bounded_source_upper,
-                );
+                    let into_ready_upper = timestamper
+                        .reclock_frontier(ready_upper.borrow())
```

Comment:
its so nice that we can just use the meet of the follower and the data here, its so cleannnnn